client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/object/ClassToSchema.java

```diff
@@ -64,38 +66,89 @@ public static void registerSchemaIfNeeded(Class<?> valueType) {
         }
 
         String name = valueType.getName();
-        String schema = ClassToSchema.getOrDeriveSchema(valueType);
-        UnitsInfoRecorder.registerNewParsedDto(name, schema);
-        ExecutionTracer.addParsedDtoName(name);
+        if (!UnitsInfoRecorder.isDtoSchemaRegister(name)){
+            List<Class<?>> embedded = new ArrayList<>();
+            String schema = ClassToSchema.getOrDeriveSchema(valueType, embedded);
+            UnitsInfoRecorder.registerNewParsedDto(name, schema);
+            ExecutionTracer.addParsedDtoName(name);
+            if (!embedded.isEmpty()){
+                embedded.forEach(ClassToSchema::registerSchemaIfNeeded);
+            }
 
+        }
+    }
+    /**
+     *
+     * @return a schema representation of the class along with schemas of its ref classes in the form
+     *      "name: { "type name": "schema", "type name": "schema", .... }"
+     */
+    public static String getOrDeriveSchemaWithItsRef(Class<?> klass){
```

Comment:
isn't the name confusing? you are always re-deriving a string for the same input `klass`. maybe should create a cache with `klass` as key?

Reply:
add more doc to explain this method